### ISSUE TITLE
AppWorld Dataset RL Training

### ISSUE THREAD

**Issue by NitinAggarwal1** [NovaSky-AI/SkyRL#72]
Hi Everyone , 

This is great work and thanks for sharing it . I would like to train the AppWorld Dataset using RL algorithms including the LOOP paper (https://arxiv.org/pdf/2502.01600)

<Here are my thoughts  > 
1. Utilise the Appworld Data set . https://appworld.dev
2. There is already an Environment for Appworld which we can setup as a server as well.
3. Utilise the skyRL system to async train on training data, using multiple gpu setting while generating multi-step trajectories for the task . 
4. Write custom reward function that works at token level or turn level as mentioned in the paper. 

Please let me know your thoughts/ suggestions on it . It might be a great addition to examples like text2sql 

Best Regards,
Nitin Aggarwal

**Comment by lynnliu030** [comment]
Hi Nitin, thanks for providing the new task! The workflow makes sense to me and you can reference documentation [here](https://skyrl.readthedocs.io/en/latest/tutorials/new_env.html) for creating a new env. Turn-level / token-level reward is not included in our current set of env examples, and thus will be great addition as well. Let us know if you have any troubles setting this up!

**Comment by tyler-griggs** [comment]
Hi @NitinAggarwal1 thanks for your interest in SkyRL, this sounds like a very cool application!

If I understand correctly, AppWorld already has an implementation of the "AppWorld Engine", which I assume is their implementation of the "Environment" and "Agent Loop". Is that right? In this case, I would actually recommend not implementing an environment in SkyRL-Gym. Instead, I would create a custom [Generator](https://github.com/NovaSky-AI/SkyRL/blob/main/skyrl-train/skyrl_train/generators/base.py) class that connects the SkyRL training stack to the AppWorld Engine environment. 

When creating a custom Generator for an existing environment (ie, AppWorld Engine), the custom Generator is typically simple and includes two elements: 
1) Passing the inference engine URL to the environment. AppWorld Engine probably talks to AppWorld Engine over HTTP, maybe via LiteLLM or OpenAI API. The custom Generator should provide AppWorld Engine whatever information is needed for the AppWorld Engine to use the inference engine that the SkyRL stack spins up.
2) Formatting AppWorld Engine output to adhere to SkyRL's [GeneratorOutput](https://github.com/NovaSky-AI/SkyRL/blob/main/skyrl-train/skyrl_train/generators/base.py#L13-L19) format. This format is what the training stack expects.

Then, the custom Generator's `generate()` method should run the AppWorld Engine on the current set of prompts. We are working on creating an example of this workflow that I will share with you once ready (in the next couple days). 

We are very happy to help you set this up, so as you're working on this feel free to post any questions here, or you can [join the discord]( https://discord.gg/AjEcBGDv) for higher throughput communication.

**Comment by NitinAggarwal1** [comment]
Thank you for the advice , I will try it .

**Comment by dipta007** [comment]
> Thank you for the advice , I will try it .

Could you do it? I am now looking for an AppWorld-RL implementation
